syntax = "proto3";

package code_talker;

enum Role {
    UNSPECIFIED_ROLE = 0;
    CODEMASTER = 1;
    INTERPRETER = 2;
}

message PlayerType {
    int32 team = 1;
    Role role = 2;
}

message AgentSet {
    repeated string codenames = 1;
}

message SecretInformation {
    map<int32, AgentSet> agent_sets = 1;
}

message CommonInformation {
    map<int32, int32> identity_counts = 1;
    map<int32, AgentSet> agent_sets = 2;
    repeated Turn turn_history = 3;
}

message Clue {
    string word = 1;
    int32 quantity = 2;
}

message SharedClue {
    int32 team = 1;
    Clue clue = 2;
}

message Action {
    string guess = 1;
}

message ActionOutcome {
    int32 identity = 1;
}

message SharedAction {
    int32 team = 1;
    Action action = 2;
    ActionOutcome action_outcome = 3;
}

message Turn {
    SharedClue clue = 1;
    repeated SharedAction actions = 2;
}

message TeamOutcomes {
    repeated int32 undetermined = 1;
    repeated int32 found_agents = 2;
    repeated int32 found_assassin = 3;
}